pkg/contract-manager/reservation_controller.go

```diff
@@ -47,13 +56,318 @@ type ReservationReconciler struct {
+			}
+		}
+		return ctrl.Result{}, nil
+	case nodecorev1alpha1.PhaseSolved:
+		klog.Infof("Reservation %s solved", reservation.Name)
+		// TODO: for Ale, create the Allocation Object
+		return ctrl.Result{}, nil
+	case nodecorev1alpha1.PhaseFailed:
+		klog.Infof("Reservation %s failed", reservation.Name)
+		return ctrl.Result{}, nil
+	}
 
 	return ctrl.Result{}, nil
 }
 
+// reserveFlavour reserves a flavour with the given flavourID
+func reserveFlavour(ctx context.Context, reservation *reservationv1alpha1.Reservation) (*Transaction, error) {
+	flavourID := reservation.Spec.FlavourID
+	body := map[string]interface{}{
+		"flavourID": flavourID,
+	}
+
+	klog.Infof("Reservation %s for flavour %s", reservation.Name, flavourID)
+
+	selectorBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the POST request to the server
+	url := discoverymanager.SERVER_ADDR + "/reserveflavour/" + flavourID
+	klog.Infof("Sending POST request to %s", url)
+	resp, err := http.Post(url, "application/json", bytes.NewBuffer(selectorBytes))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Check if the response status code is 200 (OK)
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response status code: %d", resp.StatusCode)
+	}
+
+	// Read the response body
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var transaction Transaction
+	if err := json.Unmarshal(respBody, &transaction); err != nil {
+		return nil, err
+	}
+
+	return &transaction, nil
+}
+
+// purchaseFlavour purchases a flavour with the given flavourID
+func purchaseFlavour(ctx context.Context, transaction *Transaction, reservation *reservationv1alpha1.Reservation) (*ResponsePurchase, error) {
+	flavourID := transaction.FlavourID
+	body := map[string]interface{}{
+		"transactionID": transaction.TransactionID,
+		"flavourID":     flavourID,
+		"buyerID":       reservation.Spec.Buyer.NodeID,
+	}
+
+	selectorBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the POST request to the server
+	resp, err := http.Post(discoverymanager.SERVER_ADDR+"/purchaseflavour/"+flavourID, "application/json", bytes.NewBuffer(selectorBytes))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// Check if the response status code is 200 (OK)
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response status code: %d", resp.StatusCode)
+	}
+
+	// Read the response body
+	respBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var purchase ResponsePurchase
+	if err := json.Unmarshal(respBody, &purchase); err != nil {
+		return nil, err
+	}
+
+	return &purchase, nil
+}
+
+// forgeContract creates a Contract CR
+func forgeContractCR(ctx context.Context, reservation *reservationv1alpha1.Reservation, flavour *nodecorev1alpha1.Flavour) *reservationv1alpha1.Contract {
```

Comment:
not yet needed `ctx`, change everywhere in your code, i saw it somewhere else

Reply:
Fixed everywhere.